Add Tado weather support

Noltari opened the pull request:
## Proposed change
Add weather support for the Tado integration.
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->


## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [x] New feature (which adds functionality to an existing integration)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Example entry for `configuration.yaml`:
<!--
  Supplying a configuration snippet, makes it easier for a maintainer to test
  your PR. Furthermore, for new integrations, it gives an impression of how
  the configuration would look like.
  Note: Remove this section if this PR does not have an example entry.
-->

```yaml
# Example configuration.yaml
tado:
  username: !secret tado_username
  password: !secret tado_password
```

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: https://github.com/home-assistant/home-assistant.io/pull/16096

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] The code has been formatted using Black (`black --fast homeassistant tests`)
- [x] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] Untested files have been added to `.coveragerc`.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:
<!--
  The Integration Quality Scale scores an integration on the code quality
  and user experience. Each level of the quality scale consists of a list
  of requirements. We highly recommend getting your integration scored!
-->

- [ ] No score or internal
- [ ] 🥈 Silver
- [ ] 🥇 Gold
- [ ] 🏆 Platinum

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

Noltari committed:
Tado: add weather support

Signed-off-by: Álvaro Fernández Rojas <noltari@gmail.com>

Noltari committed:
Tado: add weather tests

Signed-off-by: Álvaro Fernández Rojas <noltari@gmail.com>

Noltari committed:
tado: fix pylint warnings

Also improve init to avoid calling getMe twice.

Signed-off-by: Álvaro Fernández Rojas <noltari@gmail.com>

Noltari committed:
tests: add missing tado weather fixture

Signed-off-by: Álvaro Fernández Rojas <noltari@gmail.com>

Noltari committed:
tado: fix weather test

Signed-off-by: Álvaro Fernández Rojas <noltari@gmail.com>

hencoappel commented on homeassistant/components/tado/entity.py:
```diff
@@ -32,6 +32,26 @@ def should_poll(self):
+    def device_info(self):
+        """Return the device_info of the device."""
+        return {
+            "identifiers": {(DOMAIN, f"{self.home_id}")},
```
Any reason to use the f-string rather than just reference self.home_id directly?

Noltari replied:
Fixed on latest commit, thanks.

hencoappel commented on homeassistant/components/tado/weather.py:
```diff
@@ -0,0 +1,110 @@
+        """Initialize."""
+        self._tado = tado
+        self._data = self._tado.data["weather"]
+        super().__init__(tado)
```
Shouldn't the super constructor be called first?

Noltari replied:
Yes, should be fixed now.

hencoappel reviewed (commented):
Hey,

Left some small comments. Overall looks good though. Looking forward to the change getting in!

Noltari committed:
Tado: add some small fixes

Signed-off-by: Álvaro Fernández Rojas <noltari@gmail.com>

bouwew commented:
@Noltari I'm sorry, I don't mean to dismiss your work, but why would you want to add this to the Tado integration?
Don't you think that the company behind Tado buys the weather-data from one of the commonly available weather sources, which is probably already covered via one of the existing HA Core weather integrations?

MatthewFlamm commented:
I looked this over and it looked good to my eyes.  I did notice that you have a humidity method that seems unnecessary, but you are forced by the underlying integration to implement it.  I created a PR #45095 to address this.

Noltari commented:
> @Noltari I'm sorry, I don't mean to dismiss your work, but why would you want to add this to the Tado integration?
> Don't you think that the company behind Tado buys the weather-data from one of the commonly available weather sources, which is probably already covered via one of the existing HA Core weather integrations?

You're right, but it may be interesting to access Tado weather data, since this is used by Tado to decide if your heating is powered on or not.
BTW, I've compared it with OWM and it doesn't match exactly the data provided by them. There seems to be a slight difference in the temperature and sometimes in the condition too.

Noltari commented:
> I looked this over and it looked good to my eyes. I did notice that you have a humidity method that seems unnecessary, but you are forced by the underlying integration to implement it. I created a PR #45095 to address this.

@MatthewFlamm Yes, there's no way to get rid of humidity, since it's required by Home Assistant's Weather Entity.

bouwew commented:
@Noltari You could add the "values of interest" as sensors that are placed under the Tado hub, or whatever it's called.
This is how we show, for instance, the outdoor_temperature in the Plugwise integration:
![image](https://user-images.githubusercontent.com/11290930/104490392-15f78180-55d1-11eb-8319-8696b5bdfc04.png)

bouwew commented:
I've just noticed in the mentioned architecture PR that the provided weather data is very limited. Isn't it much simpler and quicker to add 3 sensors instead of adding a full weather implementation?

Noltari committed:
Tado: export weather data as sensors

Signed-off-by: Álvaro Fernández Rojas <noltari@gmail.com>

Noltari commented:
@bouwew I added a new commit exporting the weather data as normal sensors.

Noltari committed:
Tado: tests: replace fog by ATTR_CONDITION_FOG

Signed-off-by: Álvaro Fernández Rojas <noltari@gmail.com>

Noltari committed:
Tado: add home sensors test

Signed-off-by: Álvaro Fernández Rojas <noltari@gmail.com>

Noltari committed:
tado: sensor: set weather data on init

Signed-off-by: Álvaro Fernández Rojas <noltari@gmail.com>

elupus commented on homeassistant/components/tado/sensor.py:
```diff
@@ -68,6 +84,114 @@ async def async_setup_entry(
+                self._tado_weather_data["outsideTemperature"][
+                    "celsius" if self._is_metric else "fahrenheit"
+                ],
+                TEMP_CELSIUS if self._is_metric else TEMP_FAHRENHEIT,
```
Just let core handle the conversion. Just return the current unit and value

Noltari replied:
@elupus both units are provided, so I was just trying to avoid any conversions.
Anyway, I will change it in order to always return celsius.

elupus commented on homeassistant/components/tado/weather.py:
```diff
@@ -0,0 +1,110 @@
+    @property
+    def temperature_unit(self):
+        """Return the unit of measurement."""
+        return TEMP_CELSIUS if self._is_metric else TEMP_FAHRENHEIT
```
As above. Return current unit a d let core handle conversion.

elupus reviewed (commented):
A quick glance over and most things look okey.

Noltari committed:
tado: let core handle the temperature conversion

This change was suggested by @elupus:
https://github.com/home-assistant/core/pull/44807#discussion_r571997770

Signed-off-by: Álvaro Fernández Rojas <noltari@gmail.com>

Noltari commented:
Ping @michaelarnauts @bdraco

Noltari commented:
Rebased on current dev branch (no changes).
@bdraco can you take a look at this?

bdraco commented on homeassistant/components/tado/weather.py:
```diff
@@ -0,0 +1,107 @@
+
+    tado = hass.data[DOMAIN][entry.entry_id][DATA]
+
+    async_add_entities([TadoWeatherEntity(tado)], True)
```
```suggestion
    async_add_entities([TadoWeatherEntity(tado)])
```

Does this need to update on add?

bdraco reviewed (changes requested):
LGTM, but we need to wait for the outcome of https://github.com/home-assistant/architecture/issues/485 to decide if we can merge this.

Noltari committed:
tado: remove weather

Only sensors will be exposed until a decission regarding the lack of humidity
is made.

Signed-off-by: Álvaro Fernández Rojas <noltari@gmail.com>

Noltari commented:
> LGTM, but we need to wait for the outcome of [home-assistant/architecture#485](https://github.com/home-assistant/architecture/issues/485) to decide if we can merge this.

@bdraco instead of waiting for the outcome of https://github.com/home-assistant/architecture/issues/485, I decided to remove weather and leave the normal sensors.
I will open a new PR adding weather if the decision taken in https://github.com/home-assistant/architecture/issues/485 allows it.

bdraco commented on homeassistant/components/tado/const.py:
```diff
@@ -32,6 +32,19 @@
     PRESET_AWAY,
     PRESET_HOME,
 )
+from homeassistant.components.weather import (
```
Looks left over?

Noltari replied:
@bdraco no, it's not.
I'm using this to transform the condition reported by Tado to a home assistant standard value.
See:
https://github.com/home-assistant/core/pull/44807/files#diff-851f046c721e3d1e7fa5b0f9948981858b92bf6e623f23d923b7b6c7add24a98R51-R56
https://github.com/home-assistant/core/pull/44807/files#diff-851f046c721e3d1e7fa5b0f9948981858b92bf6e623f23d923b7b6c7add24a98R185-R191

bdraco replied:
While it does work, and I'm not specifically opposed to this, I'm not sure we should be partially implementing the weather platform on the sensor platform.  I'll ask for a second opinion on this.

MartinHjelmare replied:
No we shouldn't use weather constants for the sensor platform.

bdraco commented on homeassistant/components/tado/const.py:
```diff
@@ -48,6 +61,21 @@
 UPDATE_TRACK = "update_track"
 
+# Weather
+CONDITIONS_MAP = {
```
Same?

bdraco commented:
Given the direction of the arch issue, I'd like to get a second opinion before merging: https://github.com/home-assistant/core/pull/44807#discussion_r583686499

Noltari commented:
@bdraco rebased on latest dev branch (conflict with 198ecb0945aa062971afc471fc860b9069a567da).

Noltari committed:
tado: remove weather constants

Signed-off-by: Álvaro Fernández Rojas <noltari@gmail.com>

Noltari commented:
@bdraco @MartinHjelmare rebased on latest dev branch and removed usage of weather constants.

bdraco commented:
Will retest shortly

bdraco reviewed (approved):
LGTM

<img width="366" alt="Screen Shot 2021-03-10 at 10 31 01 AM" src="https://user-images.githubusercontent.com/663432/110693643-c1fdb900-818b-11eb-912e-5537908d0746.png">
